tensorflow/core/kernels/parse_tensor_test.cc

```diff
@@ -0,0 +1,236 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <memory>
+#include <vector>
+#include <string>
+
+#include "tensorflow/core/common_runtime/device.h"
+#include "tensorflow/core/common_runtime/device_factory.h"
+#include "tensorflow/core/framework/allocator.h"
+#include "tensorflow/core/framework/fake_input.h"
+#include "tensorflow/core/framework/node_def_builder.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/framework/types.pb.h"
+#include "tensorflow/core/kernels/ops_testutil.h"
+#include "tensorflow/core/kernels/ops_util.h"
+#include "tensorflow/core/framework/register_types.h"
+
+namespace tensorflow {
+namespace {
+
+class SerializeTensorOpTest : public OpsTestBase {
+ protected:
+  template <typename T>
+  void MakeOp() {
+    TF_ASSERT_OK(
+        NodeDefBuilder("myop", "SerializeTensor")
+            .Input(FakeInput(DataTypeToEnum<T>::value))
+            .Finalize(node_def()));
+    TF_ASSERT_OK(InitOp());
+  }
+  template <typename T>
+  void BuildParseNodeDef(NodeDef* parse) {
+    TF_ASSERT_OK(NodeDefBuilder("parse", "ParseTensor")
+                     .Input(FakeInput(DT_STRING))
+                     .Attr("out_type", DataTypeToEnum<T>::value)
+                     .Finalize(parse));
+  }
+  Tensor ParseSerialized(const NodeDef& parse_node_def, Tensor* serialized) {
+    std::unique_ptr<Device> device(
+        DeviceFactory::NewDevice("CPU", {}, "/job:a/replica:0/task:0"));
+    gtl::InlinedVector<TensorValue, 4> inputs;
+    inputs.push_back({nullptr, serialized});
+    Status status;
+    std::unique_ptr<OpKernel> op(
+        CreateOpKernel(DEVICE_CPU, device.get(),
+                       cpu_allocator(), parse_node_def,
+                       TF_GRAPH_DEF_VERSION, &status));
+    TF_EXPECT_OK(status);
+    OpKernelContext::Params params;
+    params.device = device.get();
+    params.inputs = &inputs;
+    params.frame_iter = FrameAndIter(0, 0);
+    params.op_kernel = op.get();
+    std::vector<AllocatorAttributes> attrs;
+    test::SetOutputAttrs(&params, &attrs);
+    OpKernelContext ctx(&params);
+    op->Compute(&ctx);
+    TF_EXPECT_OK(status);
+    Tensor* output = ctx.mutable_output(0);
+    return *output;
+  }
+};
+
+TEST_F(SerializeTensorOpTest, SerializeTensorOpTest_half) {
+  MakeOp<Eigen::half>();
```

Comment:
Thanks for doing this!  I like that you can see the names of the tests, which to me was the most important thing to have.

Now that I see the repetition more clearly, I think it might be fine to do the following to reduce it a little bit:

Add a function:

```
template <typename T>
void ConstructOp(std::function<T(int)> functor) {
  MakeOp<T>();
  AddInput<T>(TensorShape({10}), functor);
}
```
and another function called ParseSerializedOutput<T>() which does:

NodeDef parse;
code in BuildParseNodeDef
return ParseSerialized(parse, GetOutput(0));

So that each test looks like:

```
ConstructOp<float>([](int x) -> float { ... };);
TF_ASSERT_OK(RunOpKernel());
Tensor parse_output = ParseSerializedOutput<float>();
test::ExpectTensorEqual<float>(parse_putput, GetInput(0));
```

There's still some duplication, but I think it makes the code a bit more readable; the first statement indicates that you are building the op, the second says you are running it.  The third is all the code required to produce the Tensor that is the inverse of the output, and the last statement shows that you expect the expression to be reversible.

Let me know what you think of this!